### PR TITLE
Enhancement 117: Add in the Toolbox and tests that rely on it

### PR DESCRIPTION
**Pull request by alexowens90**

### CODE REVIEW

File: cpp/arcticdb/toolbox/python_bindings.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    using namespace arcticdb::toolbox::apy;
+    using namespace arcticdb::storage;
+
+    tools.def("print_mem_usage", &util::print_total_mem_usage);
+
+    py::class_<LibraryTool, std::shared_ptr<LibraryTool>>(tools, "LibraryTool")
+            .def(py::init<>([](std::shared_ptr<Library> lib) {
+                return std::make_shared<LibraryTool>(lib);
+            }))
+            .def("read", &LibraryTool::read)
```
Comment (mehertz) [inline cpp/arcticdb/toolbox/python_bindings.cpp:22]: I love the cleanup. Worth adding a docstring just to this method to make it clear what usage looks like? This is a super useful method IMO.

Comment (mehertz) [reply cpp/arcticdb/toolbox/python_bindings.cpp:22]: Oh I see, this is probably not the right place for a docstring. Should be in the Python wrapper.

File: cpp/arcticdb/codec/python_bindings.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+namespace arcticdb {
+
+class DynamicFieldBuffer {
+  public:
+
+    DynamicFieldBuffer(const TypeDescriptor &td, const py::buffer& data, const py::buffer& shapes) :
+        td_(td),
+        //        d_info_(data.request()),
+        //     s_info_(shapes.request()),
```
Comment (mehertz) [inline cpp/arcticdb/codec/python_bindings.cpp:30]: Remove

File: cpp/arcticdb/codec/python_bindings.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  public:
+    BufferPairDataSink() :
+        values_(std::make_shared<Buffer>()),
+        shapes_(std::make_shared<Buffer>()) {
+    }
+
+    /**
+     * @param size
+     * @return a pointer to the first byte of a contiguous memory chunk of size bytes
+     */
```
Comment (mehertz) [inline cpp/arcticdb/codec/python_bindings.cpp:127]: Wrong docstring style. Either remove or change from javadoc style

**Review by mehertz — APPROVED** [review]

File: cpp/arcticdb/codec/python_bindings.cpp (hunk 3 above)
Comment (qc00) [reply cpp/arcticdb/codec/python_bindings.cpp:127]: This seems to be Doxygen-compatible and works well in most IDEs.

File: python/arcticdb/toolbox/library_tool.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        key_type,
+                    )
+                )
+
+        return keys
+
+    def find_keys(self, key_type: KeyType) -> Union[List[AtomKey], List[RefKey]]:
+        return self.impl.find_keys(key_type)
+
+    def find_keys_for_id(self, key_type: KeyType, id: Union[str, int]) -> Union[List[AtomKey], List[RefKey]]:
```
Comment (qc00) [inline python/arcticdb/toolbox/library_tool.py:44]: Do we actually support `int` id type in the python_bindings?

Comment (qc00) [inline python/arcticdb/toolbox/library_tool.py:44]: Also, if most of the methods are pass-through, I wonder if we could just make this class extend `LibraryToolImpl`? `pybind11` should generate type-annotated Python methods by default.

Comment (alexowens90) [reply python/arcticdb/toolbox/library_tool.py:44]: Good idea, will see if it works

File: python/arcticc/pb2/__init__.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import os as _os
+import sys as _sys
+import google.protobuf as _protobuf
+
+_proto_ver = _protobuf.__version__.split(".")[0]
+if _proto_ver in "34":
+    # Use the namespace package (https://peps.python.org/pep-0420) feature to select the pb2 files matching the protobuf
+    _protos_path = _os.path.abspath(
+        _os.path.join(_os.path.dirname(__file__), "..", "..", "arcticdb", "proto", _proto_ver, "arcticc", "pb2")
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply python/arcticc/pb2/__init__.py:9]: I just rebased off Alex's branch to simplify merging if all tests passed and no other changes were required. Will rebase off master again before merging.

File: cpp/arcticdb/codec/python_bindings.cpp (hunk 3 above)
Comment (alexowens90) [reply cpp/arcticdb/codec/python_bindings.cpp]: Removed

**Pull request merged**